lib/rust/flexer/tests/flexer_generated_api_test.rs

```diff
@@ -0,0 +1,521 @@
+use lazy_reader::{BookmarkId,LazyReader,Reader};
+use flexer::group;
+
+
+// ===========
+// === AST ===
+// ===========
+
+/// A very simple AST, sufficient for the simple language being defined.
+#[derive(Clone,Debug,PartialEq)]
+pub enum AST {
+    /// A word from the input, consisting of a sequence of all `a` or all `b`.
+    Word(String),
+    /// A token that the lexer is unable to recognise.
+    Unrecognised(String)
+}
+
+
+
+// ==================
+// === Test Lexer ===
+// ==================
+
+/// The definition of a test lexer for the above-described language.
+#[derive(Debug)]
+pub struct TestLexer<Reader:LazyReader> {
+    lexer: Flexer<TestState,AST,Reader>
+}
+
+impl<Reader:LazyReader> TestLexer<Reader> {
+    /// Creates a new instance of this lexer.
+    pub fn new(reader:Reader) -> Self {
+        let lexer = Flexer::new(reader);
+        TestLexer{lexer}
+    }
+}
+
+/// Implementations of functionality used by the lexer.
+///
+/// These functions are provided by the user, by hand.
+#[allow(missing_docs)]
+impl<Reader:LazyReader> TestLexer<Reader> {
+    pub fn on_first_word(&mut self) {
+        let str = self.current_match.clone();
+        let ast = AST::Word(str);
+        self.output.push(ast);
+        let id = self.seen_first_word_state;
+        self.push_state(id);
+    }
+
+    pub fn on_spaced_word(&mut self) {
+        let str = self.current_match.clone();
+        let ast = AST::Word(String::from(str.trim()));
+        self.output.push(ast);
+    }
+
+    pub fn on_err_suffix_first_word(&mut self) {
+        let ast = AST::Unrecognised(self.current_match.clone());
+        self.output.push(ast);
+    }
+
+    pub fn on_err_suffix(&mut self) {
+        self.on_err_suffix_first_word();
+        self.pop_state();
+    }
+
+    pub fn on_no_err_suffix_first_word(&mut self) {}
+
+    pub fn on_no_err_suffix(&mut self) {
+        self.on_no_err_suffix_first_word();
+        self.pop_state();
+    }
+}
+
+impl<Reader:LazyReader> Deref for TestLexer<Reader> {
+    type Target = Flexer<TestState,AST,Reader>;
+    fn deref(&self) -> &Self::Target {
+        &self.lexer
+    }
+}
+
+impl<Reader:LazyReader> DerefMut for TestLexer<Reader> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.lexer
+    }
+}
+
+/// Generated functionality used at runtime by the lexer.
+#[allow(missing_docs)]
+impl<Reader:LazyReader> TestLexer<Reader> {
+
+    /// Executes the lexer on the input provided by the reader, resulting in a
+    /// series of tokens.
+    pub fn run(&mut self) -> Result<AST> {
+        self.reader.advance_char();
+
+        while self.gen_run_current_state() == StageStatus::ExitSuccess {}
+
+            match self.status {
+                StageStatus::ExitFinished => Result::Success(self.result().clone()),
+                StageStatus::ExitFail     => Result::Failure(self.result().clone()),
+                _                         => Result::Partial(self.result().clone())
+            }
+    }
```

Comment:
I don't like cloning here. Use `mem::replace` instead to get the result instead :)